src/test/java/com/github/bibsysdev/nva/cristin/person/CristinPersonsHandlerTest.java

```diff
@@ -0,0 +1,227 @@
+        assertTrue(response.getBody().contains(LANGUAGE_INVALID_ERROR_MESSAGE));
+    }
+
+    @Test
+    public void cristinApiClientWillStillGenerateQueryPersonsUrlEvenWithoutParameters()
+        throws IOException, URISyntaxException {
+        cristinApiClientStub.generateQueryPersonsUrl(null);
+    }
+
+    @Test
+    public void readerThrowsIoExceptionWhenReadingInvalidJson() {
+        InputStream inputStream = new ByteArrayInputStream(INVALID_JSON.getBytes(StandardCharsets.UTF_8));
+        InputStreamReader reader = new InputStreamReader(inputStream);
+        Executable action = () -> CristinApiClient.fromJson(reader, CristinPerson.class);
+        assertThrows(IOException.class, action);
+    }
+
+    @BeforeEach
+    void setUp() {
+        cristinApiClientStub = new CristinApiClientStub();
+        context = mock(Context.class);
+        output = new ByteArrayOutputStream();
+        handler = new CristinPersonsHandler(cristinApiClientStub, environment);
+    }
+
+    @ParameterizedTest
+    @ArgumentsSource(TestPairProvider.class)
+    void handlerReturnsExpectedBodyWhenRequestInputIsValid(String expected) throws IOException {
+        var actual = sendDefaultQuery().getBody();
+        assertEquals(objectMapper.readTree(expected), objectMapper.readTree(actual));
+    }
+
+    @Test
+    void handlerReturnsNonEnrichedBodyWhenEnrichingFails() throws Exception {
+        cristinApiClientStub = Mockito.spy(cristinApiClientStub);
+        doThrow(new IOException()).when(cristinApiClientStub).getPerson(any(), any());
+        handler = new CristinPersonsHandler(cristinApiClientStub, environment);
+        GatewayResponse<PersonsWrapper> response = sendDefaultQuery();
+        var expected = getReader(NVA_PERSONS_NON_ENRICHED_JSON_FILE);
+
+        assertEquals(objectMapper.readTree(expected), objectMapper.readTree(response.getBody()));
+    }
+
+    @Test
+    void returnNvaPersonWhenCallingNvaPersonBuilderMethodWithValidCristinPerson() throws Exception {
+        var expected = getReader(NVA_ONE_PERSON_JSON_FILE);
+        var cristinGetPerson = getReader(CRISTIN_ONE_PERSON_JSON_FILE);
+        CristinPerson cristinPerson = attempt(
+            () -> objectMapper.readValue(cristinGetPerson, CristinPerson.class)).get();
+        NvaPerson nvaPerson = new NvaPersonBuilder(cristinPerson).build();
+        nvaPerson.setContext(PERSON_LOOKUP_CONTEXT_URL);
+        var actual = attempt(() -> objectMapper.writeValueAsString(nvaPerson)).get();
+
+        assertEquals(objectMapper.readTree(expected), objectMapper.readTree(actual));
+    }
+
+    @Test
+    void handlerReturnsPersonsWrapperWithAllMetadataButEmptyHitsArrayWhenNoMatchesAreFoundInCristin() throws Exception {
+        cristinApiClientStub = Mockito.spy(cristinApiClientStub);
+        var emptyArray = new InputStreamReader(IoUtils.stringToStream(EMPTY_LIST_STRING), Charsets.UTF_8);
+        doReturn(emptyArray).when(cristinApiClientStub).getQueryResponse(any());
```

Comment:
when() should be used instead of doReturn(); might be a moot point when wiremock is used.

Reply:
I think it has to do with if we are using spy or mock, that `doReturn` should be used when spying on actual objects, while `when` is used on mocked classes. I'm not 100% sure but i think is was because of that

Reply:
doReturn should only be used when spying real objects and calling it's methods has side-effects. It's not meant to generally be used when spying real objects, when() is still the general use case for that. It can also be used to override previous stubbing of exceptions.